File: ai-service-example/ai_implementation/AimlApiLLM.py
import requests
import time
import os
class AimlApiLLM:
    def __init__(self, api_key: str, model_name: str = "o1-preview"):
        self.api_key = api_key
        self.model_name = model_name

    def generate_response(self, prompt: str) -> str:
        url = "https://api.aimlapi.com/chat/completions"
        retries = 5
        
        for attempt in range(retries):
            response = requests.post(
                url=url,
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": self.model_name,
                    "messages": [
                        {"role": "user", "content": prompt}
                    ],
                    "max_tokens": 512,
                    "stream": False,
                },
            )

            try:
                response.raise_for_status()  # Raise an error for bad responses
                return response.json()["choices"][0]["message"]["content"]  # Extract response content
            except requests.exceptions.HTTPError as e:
                if response.status_code == 429:
                    # Too many requests; wait and retry
                    wait_time = 2 ** attempt  # Exponential backoff
                    print(f"Rate limit exceeded. Waiting for {wait_time} seconds before retrying...")
                    time.sleep(wait_time)
                else:
                    # Raise other HTTP errors
                    raise e
        
        raise Exception("Exceeded maximum retries due to rate limits.")

if __name__ == "__main__":
    api_key = "1d525acd15344646a522ba71620f23df"  # Replace with your actual API key
    aiml_llm = AimlApiLLM(api_key=api_key)
    
    prompt = "What kind of model are you?"
    response = aiml_llm.generate_response(prompt)
    print("Generated Response:", response)

File: backend/route/aimlApiLLM.js
const axios = require('axios');
const csv = require('csv-parser');
const fs = require('fs');
const readlineSync = require('readline-sync');
const ragcy = require('ragcy');

const API_KEY = process.env.RAGCY;
const client = ragcy.init({ key: API_KEY });

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

class AimlApiLLM {
  constructor(apiKey, modelName = 'o1-mini') {
    this.apiKey = apiKey;
    this.modelName = modelName;
  }

  async readCSVData(filePath) {
    return new Promise((resolve, reject) => {
      let data = '';
      fs.createReadStream(filePath)
        .pipe(csv())
        .on('data', (row) => {
          data += JSON.stringify(row) + '\n';
        })
        .on('end', () => resolve(data))
        .on('error', reject);
    });
  }

  async getRagcySuggestion(prompt, corpusId) {
    try {
      const response = await client.getSuggestions({
        corpusId: corpusId,
        query: prompt,
      });

      return response.suggestions[0].text;
    } catch (error) {
      console.error('Error fetching Ragcy suggestion:', error);
      return '';
    }
  }

  async generateResponse(prompt, previousPrompt) {
    const url = 'https://api.aimlapi.com/chat/completions';
    const retries = 5;

    const systemPrompt = `You are a doctor assistant chatbot.`;

    const combinedPrompt = `Previous Prompt: ${systemPrompt}\nUser Prompt: ${prompt}`;

    // const ragcySuggestion = await this.getRagcySuggestion(prompt, 'your_corpus_id_here');

    const diseaseMedicineData = await this.readCSVData('./data/disease_medicine.csv');
    const diseaseSymptomsData = await this.readCSVData('./data/disease_symptoms.csv');
    const patientData = await this.readCSVData('./data/patient_data.csv');

    const ragPrompt = `
      Disease and Medicine Information: ${diseaseMedicineData}
      Disease Symptoms Information: ${diseaseSymptomsData}
      Patient Data: ${patientData}
      Ragcy Suggestion: ${ragcySuggestion}
    `;
    
    for (let attempt = 0; attempt < retries; attempt++) {
      try {
        const response = await axios.post(
          url,
          {
            model: this.modelName,
            messages: [
              {
                role: "user",
                content: combinedPrompt,
              },
            ],
            max_tokens: 256,
            stream: false,
          },
          {
            headers: {
              Authorization: `Bearer ${this.apiKey}`,
              'Content-Type': 'application/json',
            },
          }
        );
        if (response.data.choices && response.data.choices[0].message) {
          return {
            responseChatbot: response.data.choices[0].message.content,
            suggestRagcy: 'Would you like to ask Ragcy for a suggestion?',
          };
        } else {
          throw new Error('Invalid response structure');
        }
      } catch (error) {
        console.error('Error during API call:', error.message);
        if (attempt === retries - 1) {
          throw new Error('Exceeded maximum retries due to API call failure.');
        }
        await sleep(2000);
      }
    }
  }
}

module.exports = AimlApiLLM;
